subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy

```diff
@@ -802,4 +802,124 @@ abstract class AbstractJavaGroovyCompileAvoidanceIntegrationSpec extends Abstrac
         then:
         failure.assertHasCause('Compilation failed; see the compiler error output for details.')
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/20398")
+    def "doesn't recompile when private element of implementation class changes and there are malformed classes on the classpath"() {
+        given:
+        buildFile << """
+            project(':b') {
+                dependencies {
+                    implementation project(':a')
+                }
+            }
+
+            project(':a') {
+                compileJava {
+                    doLast {
+                        // Create a poison class file on the classpath
+                        destinationDirectory.file('Poison.class').get().asFile.text = "foo"
+                    }
+                }
+            }
+        """
+        def sourceFile = file("a/src/main/${language.name}/ToolImpl.${language.name}")
+        sourceFile << """
+            public class ToolImpl {
+                private String thing() { return null; }
+                private ToolImpl t = this;
+            }
+        """
+        file("b/src/main/${language.name}/Main.${language.name}") << """
+            public class Main { ToolImpl t = new ToolImpl(); }
+        """
+
+        when:
+        succeeds ":b:${language.compileTaskName}"
+
+        then:
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        executedAndNotSkipped ":b:${language.compileTaskName}"
+
+        when:
+        // change signatures
+        sourceFile.text = """
+            public class ToolImpl {
+                private Number thing() { return null; }
+                private Object t = this;
+            }
+        """
+
+        then:
+        succeeds ":b:${language.compileTaskName}"
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        skipped ":b:${language.compileTaskName}"
```

Comment:
Should these be checking `:a` instead of `:b`? Or maybe both?

Reply:
In this case, a has changed a class file in an ABI-compatible way (change to the signature of a private method) so we expect a to recompile, but b to _not_ recompile, even though a is now slightly different.  Project a also has a poison class file in its outputs, though.  Previously, this would cause the hashing to blow up, fallback to hashing the entire jar, and then cause b to recompile because of the changed private method.  Now, it catches the poison class file, hashes only that class in its entirety and then continues hashing the ABI of the other non-poison class files.  The hash of a's classes therefore does not change and b is skipped as we expect.

I think the conditions above are correct, but it's easy to get confused.  Let me know if I'm missing something.

Reply:
I think I got mixed up while reviewing, yea. This looks good on second pass.